ckanext/datastore/backend/postgres.py

```diff
@@ -369,24 +386,31 @@ def _validate_record(record: Any, num: int, field_names: Iterable[str]):
 
 def _where_clauses(
         data_dict: dict[str, Any], fields_types: dict[str, Any]
-) -> list[Any]:
+) -> WhereClauses:
     filters = data_dict.get('filters', {})
-    clauses: list[Any] = []
+    clauses: WhereClauses = []
+
+    idx_gen = itertools.count()
 
     for field, value in filters.items():
         if field not in fields_types:
             continue
         field_array_type = _is_array_type(fields_types[field])
-        # "%" needs to be escaped as "%%" in any query going to
-        # connection.execute, otherwise it will think the "%" is for
-        # substituting a bind parameter
-        field = field.replace('%', '%%')
+
         if isinstance(value, list) and not field_array_type:
-            clause_str = (u'"{0}" in ({1})'.format(field,
-                          ','.join(['%s'] * len(value))))
-            clause = (clause_str,) + tuple(value)
+            placeholders = [
+                f"value_{next(idx_gen)}" for _ in value
+            ]
+            clause_str = ('"{0}" in ({1})'.format(
+                field,
+                ','.join(f":{p}" for p in placeholders)
+            ))
```

Comment:
while making changes here let's make this sql generation from fieldnames properly safe. I would use my old `identifier` function but maybe there's a `sa.column` or something that is better now?

Reply:
Yes, `sa.column` solves the issue. Is it worth backporting? every field itself is checked in the beginning of the method against a list of existing column names: `field not in fields_types:`.

The only edge case I can imagine - datastore field contains `"` in its name. I doubt that there is a possible vector for SQL ingestion, because all statements that rely on where-clauses are executed as a single query. So basically user who uploads a "dangerous" resource can only affect this exact resource's datastore table.